### PR TITLE
Add binary compatibility test [API-1683]

### PR DESCRIPTION
**Pull request by mehmettokgoz**
Closes https://github.com/hazelcast/hazelcast-go-client/issues/902.

### CODE REVIEW

**Review by yuce — CHANGES_REQUESTED** [review]
Thanks. Here are some initial ideas. I'll continue to review...

File: internal/serialization/compatibility/binary_compatability_test.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/hazelcast/hazelcast-go-client/internal/serialization"
+	pubserialization "github.com/hazelcast/hazelcast-go-client/serialization"
+)
+
+var (
+	dataMap    = make(map[string]serialization.Data)
```
[the remark this answers is not in the document]
Comment (mehmettokgoz) [reply internal/serialization/compatibility/binary_compatability_test.go]: Done at https://github.com/hazelcast/hazelcast-go-client/pull/908/commits/d42b58a0e824c0f5be0b94c35cd7166931c29339.

File: internal/serialization/compatibility/binary_compatability_test.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+					require.Equal(t, obj, readObject)
+				})
+			}
+		}
+	}
+}
+
+func readBinaryFile(t *testing.T) {
+	for _, v := range versions {
+		b, err := ioutil.ReadFile(createFileName(v))
```
[the remark this answers is not in the document]
Comment (mehmettokgoz) [reply internal/serialization/compatibility/binary_compatability_test.go]: Done at https://github.com/hazelcast/hazelcast-go-client/pull/908/commits/d42b58a0e824c0f5be0b94c35cd7166931c29339.

File: internal/serialization/compatibility/binary_compatability_test.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+			n := i.ReadInt32()
+			if n != -1 {
+				bytes = make([]byte, n)
+				for j := int32(0); j < n; j++ {
+					bytes[j] = i.ReadByte()
+				}
+				dataMap[key] = bytes
+			} else {
+				dataMap[key] = nil
+			}
```
[the remark this answers is not in the document]
Comment (mehmettokgoz) [reply internal/serialization/compatibility/binary_compatability_test.go]: Done at https://github.com/hazelcast/hazelcast-go-client/pull/908/commits/d42b58a0e824c0f5be0b94c35cd7166931c29339.

File: internal/serialization/compatibility/binary_compatability_test.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+	err = cd.AddInt32Field("i")
+	require.NoError(t, err)
+	err = cd.AddFloat32Field("f")
+	require.NoError(t, err)
+	cfg.SetClassDefinitions(cd)
+	cfg.SetPortableFactories(&PortableFactory{})
+	cfg.SetIdentifiedDataSerializableFactories(&IdentifiedFactory{})
+	if bo == binary.LittleEndian {
+		cfg.LittleEndian = true
+	}
```
[the remark this answers is not in the document]
Comment (mehmettokgoz) [reply internal/serialization/compatibility/binary_compatability_test.go]: Converted to `cfg.LittleEndian = bo == binary.LittleEndian`

File: internal/serialization/compatibility/reference_object.go [hunk 5]
```diff
@@ -0,0 +1,796 @@
...
+	aBoolean     bool            = true
+	aByte        byte            = 113
+	aChar        uint16          = 'x'
+	aDouble      float64         = -897543.3678909
+	aShort       int16           = -500
+	aFloat       float32         = 900.5678
+	anInt        int32           = 56789
+	aLong        int64           = -50992225
+	anSqlString  string          = "this > 5 AND this < 100"
+	aString      string          = anSqlString
```
[the remark this answers is not in the document]
Comment (mehmettokgoz) [reply internal/serialization/compatibility/reference_object.go]: Removed at https://github.com/hazelcast/hazelcast-go-client/pull/908/commits/d42b58a0e824c0f5be0b94c35cd7166931c29339.

File: internal/serialization/compatibility/reference_object.go [hunk 6]
```diff
@@ -0,0 +1,796 @@
...
+		(types.OffsetDateTime)(time.Date(1923, 4, 23, 15, 44, 39, 192837465, time.FixedZone("", 5*60*60))),
+		(types.OffsetDateTime)(time.Date(2021, 6, 28, 18, 30, 55, 567891234, time.FixedZone("", -10*60*60))),
+	}
+
+	aBigInteger = big.NewInt(1314432323232411)
+	aClass      = "java.math.BigDecimal"
+	aBigDecimal = types.NewDecimal(big.NewInt(31231), 0)
+	decimals    = []types.Decimal{aBigDecimal, aBigDecimal, aBigDecimal}
+
+	nonNilList = []interface{}{aBoolean, aByte, aChar, aDouble, aShort, aFloat, anInt, aLong, aSmallString,
```
[the remark this answers is not in the document]
Comment (mehmettokgoz) [reply internal/serialization/compatibility/reference_object.go]: Removed at https://github.com/hazelcast/hazelcast-go-client/pull/908/commits/d42b58a0e824c0f5be0b94c35cd7166931c29339.

File: internal/serialization/compatibility/reference_object.go [hunk 7]
```diff
@@ -0,0 +1,796 @@
...
+		anInnerPortable, booleans, bytes, chars, doubles, shorts, floats, ints, longs, strings, aCustomStreamSerializable,
+		aCustomByteArraySerializable, anIdentifiedDataSerializable, aPortable, aDate, aLocalDate, aLocalTime, aLocalDateTime,
+		anOffsetDateTime, aBigInteger, aBigDecimal, aClass}
+)
+
+func initializeVariables() {
```
[the remark this answers is not in the document]
Comment (mehmettokgoz) [reply internal/serialization/compatibility/reference_object.go]: Refactored at https://github.com/hazelcast/hazelcast-go-client/pull/908/commits/d42b58a0e824c0f5be0b94c35cd7166931c29339.

File: internal/serialization/compatibility/reference_object.go [hunk 8]
```diff
@@ -0,0 +1,796 @@
...
+	return PortableFactoryId
+}
+
+func (p PortableFactory) Create(classID int32) serialization.Portable {
+	if classID == InnerPortableClassId {
+		return &AnInnerPortable{}
+	} else if classID == PortableClassId {
+		return &APortable{}
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (mehmettokgoz) [reply internal/serialization/compatibility/reference_object.go]: Refactored at https://github.com/hazelcast/hazelcast-go-client/pull/908/commits/d42b58a0e824c0f5be0b94c35cd7166931c29339.

File: internal/serialization/compatibility/reference_object.go [hunk 9]
```diff
@@ -0,0 +1,796 @@
...
+
+func readRawBytes(i serialization.DataInput, size int) []byte {
+	a := make([]byte, size)
+	for idx, _ := range a {
+		a[idx] = i.ReadByte()
+	}
+	return a
+}
+
+var (
```
[the remark this answers is not in the document]
Comment (mehmettokgoz) [reply internal/serialization/compatibility/reference_object.go]: Refactored at https://github.com/hazelcast/hazelcast-go-client/pull/908/commits/d42b58a0e824c0f5be0b94c35cd7166931c29339.

File: internal/serialization/compatibility/reference_object.go [hunk 10]
```diff
@@ -0,0 +1,796 @@
...
+	PortableFactoryId                   = 1
+	PortableClassId                     = 1
+	InnerPortableClassId                = 2
+	IdentifiedDataSerializableClassId   = 1
+	CustomStreamSerializableId          = 1
+	CustomByteArraySerializableId       = 2
+)
+
+var (
+	allTestObjects map[string]interface{} = nil
```
[the remark this answers is not in the document]
Comment (mehmettokgoz) [reply internal/serialization/compatibility/reference_object.go]: Refactored at https://github.com/hazelcast/hazelcast-go-client/pull/908/commits/d42b58a0e824c0f5be0b94c35cd7166931c29339.

File: internal/serialization/compatibility/reference_object.go [hunk 11]
```diff
@@ -0,0 +1,796 @@
...
+)
+
+const (
+	IdentifiedDataSerializableFactoryId = 1
+	PortableFactoryId                   = 1
+	PortableClassId                     = 1
+	InnerPortableClassId                = 2
+	IdentifiedDataSerializableClassId   = 1
+	CustomStreamSerializableId          = 1
+	CustomByteArraySerializableId       = 2
```
[the remark this answers is not in the document]
Comment (mehmettokgoz) [reply internal/serialization/compatibility/reference_object.go]: Refactored at https://github.com/hazelcast/hazelcast-go-client/pull/908/commits/d42b58a0e824c0f5be0b94c35cd7166931c29339.

File: internal/serialization/compatibility/binary_compatability_test.go [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+	s := []string{
+		"Class",
+	}
+	for _, v := range s {
+		if v == objType {
+			return true
+		}
+	}
+	return false
+}
```
[the remark this answers is not in the document]
Comment (mehmettokgoz) [reply internal/serialization/compatibility/binary_compatability_test.go]: This code was aiming to have more than one skipped type, but only `Class` is needed now. Refactored at https://github.com/hazelcast/hazelcast-go-client/pull/908/commits/d42b58a0e824c0f5be0b94c35cd7166931c29339.

File: internal/serialization/compatibility/binary_compatability_test.go [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+	}
+	return fmt.Sprintf("%d-%s-%s", version, name, byteOrderString)
+}
+
+func createFileName(version int) string {
+	return fmt.Sprintf("%d.serialization.compatibility.binary", version)
+}
+
+func skipOnDeserialize(objType string) bool {
+	p, _ := regexp.Compile("^.*(Predicate|Aggregator|Projection)$")
```
[the remark this answers is not in the document]
Comment (mehmettokgoz) [reply internal/serialization/compatibility/binary_compatability_test.go]: Added https://github.com/hazelcast/hazelcast-go-client/pull/908/commits/d42b58a0e824c0f5be0b94c35cd7166931c29339.

File: internal/serialization/compatibility/binary_compatability_test.go [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+	return s
+}
+
+func createObjectKey(name string, byteOrder binary.ByteOrder, version int) string {
+	byteOrderString := ""
+	if byteOrder == binary.BigEndian {
+		byteOrderString = "BIG_ENDIAN"
+	} else {
+		byteOrderString = "LITTLE_ENDIAN"
+	}
```
[the remark this answers is not in the document]
Comment (mehmettokgoz) [reply internal/serialization/compatibility/binary_compatability_test.go]: Refactored at https://github.com/hazelcast/hazelcast-go-client/pull/908/commits/bee5a0d60b0a18bd2cb2d35e2e1462c428820c1c.

File: internal/serialization/compatibility/binary_compatability_test.go [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+	return fmt.Sprintf("%d-%s-%s", version, name, byteOrderString)
+}
+
+func createFileName(version int) string {
+	return fmt.Sprintf("%d.serialization.compatibility.binary", version)
+}
+
+func skipOnDeserialize(t *testing.T, objType string) bool {
+	p, err := regexp.Compile("^.*(Predicate|Aggregator|Projection)$")
```
Comment (yuce) [inline internal/serialization/compatibility/binary_compatability_test.go:120]: I think it's better to make this function simpler
```
return strings.HasSuffix(o, "Predicate") || strings.HasSuffix(o, '"Aggregator") || strings.HasSuffix(o, "Predicate")
```

**Review by yuce — CHANGES_REQUESTED** [review]
One minor idea, after that we can merge.

File: internal/serialization/compatibility/binary_compatability_test.go (hunk 15 above)
Comment (mehmettokgoz) [reply internal/serialization/compatibility/binary_compatability_test.go]: Makes sense, fixed at https://github.com/hazelcast/hazelcast-go-client/pull/908/commits/2a0919c1c7bca0f09e5c43b215df75ea31a33bc5.

**Review by yuce — APPROVED** [review]
Looks good!

**Pull request merged**